File: Data_Structures_and_Algorithms/Hanoi_Towers/python/hanoi_recursion.py
def hanoi(n, source, auxiliary, destination):
    global s
    if n == 1:
        print(source, "-->", destination)
        s += 1
    else:
        hanoi(n - 1, source, destination, auxiliary)
        print(source, "-->", destination)
        s += 1
        hanoi(n - 1, auxiliary, source, destination)


if __name__ == "__main__":
    print("Enter number of disks: ")
    n = int(input())
    print()
    s = 0
    hanoi(n, "S", "A", "D")
    print("\nNumber of steps:", s)

File: Data_Structures_and_Algorithms/Hanoi_Towers/TypeScript/hanoi_recursion.ts
import * as readline from "readline";

let s: number;

function hanoi(
  n: number,
  source: string,
  auxiliary: string,
  destination: string
): void {
  if (n === 1) {
    console.log(`${source} --> ${destination}`);
    s++;
  } else {
    hanoi(n - 1, source, destination, auxiliary);
    console.log(`${source} --> ${destination}`);
    s++;
    hanoi(n - 1, auxiliary, source, destination);
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.question("Enter number of disks: ", (answer) => {
  const n: number = parseInt(answer);
  console.log();
  s = 0;
  hanoi(n, "S", "A", "D");
  console.log("\nNumber of steps:", s);
  rl.close();
});
